File: packages/json/llm-json-stream-ts/packages/llm-json-stream/src/utilities/stream_text_in_chunks.ts
/**
 * Utility function for streaming text in chunks.
 * Primarily used for testing to simulate LLM streaming behavior.
 */

export interface StreamTextOptions {
    text: string;
    chunkSize: number;
    interval: number; // milliseconds
}

/**
 * Creates an async iterable that emits text in chunks.
 * Useful for testing and simulating LLM streaming responses.
 * This uses async iterables for maximum cross-platform compatibility.
 */
export async function* streamTextInChunks(
    options: StreamTextOptions,
): AsyncGenerator<string, void, unknown> {
    const { text, chunkSize, interval } = options;

    for (let i = 0; i < text.length; i += chunkSize) {
        const chunk = text.slice(i, i + chunkSize);

        if (interval > 0 && i > 0) {
            // Wait before emitting (except for the first chunk)
            await new Promise((resolve) => setTimeout(resolve, interval));
        }

        yield chunk;
    }
}

File: packages/json/llm-json-stream-python/src/llm_json_stream/stream_text_in_chunks.py
"""Utility for streaming text in chunks (testing helper)."""

from __future__ import annotations

import asyncio
from typing import AsyncIterator


async def stream_text_in_chunks(
    text: str, chunk_size: int = 10, interval: int = 0
) -> AsyncIterator[str]:
    for i in range(0, len(text), chunk_size):
        chunk = text[i : i + chunk_size]
        if interval > 0 and i > 0:
            await asyncio.sleep(interval / 1000.0)
        yield chunk
